File: src/shared/api/researchAgent.ts
import { invokeCommand } from "./core";
import { executeWorkflowStart } from "./agent";
import type {
  ClaimEvidenceAssessment,
  AgentResourceLock,
  EvidencePacket,
  ResearchChatMigrationResult,
  ResearchChangeCheckpoint,
  ResearchChangeCheckpointUndoResult,
  ResearchChatStore,
  ResearchAgentRun,
  ResearchCapabilityDescriptor,
  ResearchPlanApproval,
  ResearchPlanExecutionAccepted,
  ResearchPlanVersion,
  ResearchTask,
  ResearchRunRecoveryResponse,
  ResearchNetworkPolicy,
  ResearchFulltextDocument,
  ResearchQuerySnapshot,
  ResearchReviewProtocol,
  ResearchReviewWorkspace,
  ResearchScreeningRecord,
  ResearchUiCommand,
  ResearchWorkspaceSnapshot,
} from "../types/researchAgent";

export const RESEARCH_RUN_CHANGED_EVENT = "latotex.research.run.changed";

function emitResearchRunChanged(projectId: string) {
  if (typeof window === "undefined" || typeof CustomEvent === "undefined") {
    return;
  }
  window.dispatchEvent(new CustomEvent(RESEARCH_RUN_CHANGED_EVENT, { detail: { projectId } }));
}

export function getResearchWorkspace(projectId: string): Promise<ResearchWorkspaceSnapshot> {
  return invokeCommand<ResearchWorkspaceSnapshot>("research_workspace_get", {
    input: { projectId },
  });
}

export function createResearchTask(
  projectId: string,
  goal: string,
  chatSessionId?: string | null,
): Promise<ResearchTask> {
  return invokeCommand<ResearchTask>("research_task_create", {
    input: { projectId, goal, chatSessionId: chatSessionId ?? null },
  });
}

export function getResearchNetworkPolicy(projectId: string): Promise<ResearchNetworkPolicy> {
  return invokeCommand<ResearchNetworkPolicy>("research_network_policy_get", {
    input: { projectId },
  });
}

export function updateResearchNetworkPolicy(input: Omit<ResearchNetworkPolicy, "updatedAt">) {
  return invokeCommand<ResearchNetworkPolicy>("research_network_policy_update", { input });
}

export function startResearchPlanningWorkflow(input: {
  projectId: string;
  taskId: string;
  prompt: string;
  modelOverride?: string;
}) {
  return executeWorkflowStart({
    projectId: input.projectId,
    workflowId: "research-plan-discussion",
    callsite: "research.workbench",
    prompt: input.prompt,
    contextRefs: [],
    modelOverride: input.modelOverride,
    bypassCache: true,
    teamMode: "off",
    harnessProfileId: "research.planning",
    profileId: "builtin-planner",
    researchTaskId: input.taskId,
  });
}

export function saveResearchPlan(input: {
  projectId: string;
  taskId: string;
  sourceMessage: string;
  authorizedProjectIds: string[];
  title?: string;
  summary?: string;
  assumptions?: string[];
  expectedArtifacts?: string[];
  acceptanceCriteria?: string[];
  steps: Array<{
    id?: string;
    enabled: boolean;
    dependencies: string[];
    capability: string;
    input: unknown;
    riskLevel: "read" | "write" | "high";
  }>;
}): Promise<ResearchPlanVersion> {
  return invokeCommand<ResearchPlanVersion>("research_plan_save", { input });
}

export function approveResearchPlan(
  projectId: string,
  taskId: string,
  version: number,
): Promise<ResearchPlanVersion> {
  return invokeCommand<ResearchPlanVersion>("research_plan_approve", {
    input: { projectId, taskId, version },
  });
}

export function getResearchChatStore(projectId: string): Promise<ResearchChatStore> {
  return invokeCommand<ResearchChatStore>("research_chat_store_get", {
    input: { projectId },
  });
}

export function replaceResearchChatStore(
  projectId: string,
  store: ResearchChatStore,
): Promise<ResearchChatStore> {
  return invokeCommand<ResearchChatStore>("research_chat_store_replace", {
    input: { projectId, store },
  });
}

export function migrateResearchChatStore(
  projectId: string,
  store: ResearchChatStore,
): Promise<ResearchChatMigrationResult> {
  return invokeCommand<ResearchChatMigrationResult>("research_chat_store_migrate", {
    input: { projectId, migrationId: "localstorage-chat-v1", store },
  });
}

export function getResearchCapabilityRegistry(): Promise<ResearchCapabilityDescriptor[]> {
  return invokeCommand<ResearchCapabilityDescriptor[]>("research_capability_registry");
}

export async function executeResearchPlan(
  projectId: string,
  taskId: string,
  version: number,
): Promise<ResearchPlanExecutionAccepted> {
  const accepted = await invokeCommand<ResearchPlanExecutionAccepted>("research_plan_execute", {
    input: { projectId, taskId, version },
  });
  emitResearchRunChanged(projectId);
  return accepted;
}

export function listResearchRuns(
  projectId?: string | null,
  includeTerminal = false,
): Promise<ResearchAgentRun[]> {
  return invokeCommand<ResearchAgentRun[]>("research_run_list", {
    input: { projectId: projectId ?? null, includeTerminal },
  });
}

export function listResearchUiCommands(projectId: string): Promise<ResearchUiCommand[]> {
  return invokeCommand<ResearchUiCommand[]>("research_ui_command_list", {
    input: { projectId },
  });
}

export async function resolveResearchUiCommand(input: {
  projectId: string;
  runId: string;
  stepId: string;
  status: "completed" | "failed";
  result?: unknown;
  diagnosticCode?: string | null;
}): Promise<ResearchPlanExecutionAccepted> {
  const accepted = await invokeCommand<ResearchPlanExecutionAccepted>("research_ui_command_resolve", { input });
  emitResearchRunChanged(input.projectId);
  return accepted;
}

export async function pauseResearchRun(projectId: string, runId: string) {
  const accepted = await invokeCommand<ResearchPlanExecutionAccepted>("research_run_pause", {
    input: { projectId, runId },
  });
  emitResearchRunChanged(projectId);
  return accepted;
}

export async function resumeResearchRun(projectId: string, runId: string) {
  const accepted = await invokeCommand<ResearchPlanExecutionAccepted>("research_run_resume", {
    input: { projectId, runId },
  });
  emitResearchRunChanged(projectId);
  return accepted;
}

export async function cancelResearchRun(projectId: string, runId: string) {
  const accepted = await invokeCommand<ResearchPlanExecutionAccepted>("research_run_cancel", {
    input: { projectId, runId },
  });
  emitResearchRunChanged(projectId);
  return accepted;
}

export async function recoverResearchRuns(projectId: string): Promise<ResearchRunRecoveryResponse> {
  const result = await invokeCommand<ResearchRunRecoveryResponse>("research_runs_recover", {
    input: { projectId },
  });
  emitResearchRunChanged(projectId);
  return result;
}

export function listResearchChangeCheckpoints(
  projectId: string,
  runId?: string | null,
): Promise<ResearchChangeCheckpoint[]> {
  return invokeCommand<ResearchChangeCheckpoint[]>("research_change_checkpoint_list", {
    input: { projectId, runId: runId ?? null },
  });
}

export async function undoResearchChangeCheckpoint(
  projectId: string,
  checkpointId: string,
): Promise<ResearchChangeCheckpointUndoResult> {
  const result = await invokeCommand<ResearchChangeCheckpointUndoResult>(
    "research_change_checkpoint_undo",
    { input: { projectId, checkpointId } },
  );
  emitResearchRunChanged(projectId);
  return result;
}

export function listResearchPlanApprovals(projectId: string): Promise<ResearchPlanApproval[]> {
  return invokeCommand<ResearchPlanApproval[]>("research_plan_approval_list", {
    input: { projectId },
  });
}

export function listResearchResourceLocks(projectId: string): Promise<AgentResourceLock[]> {
  return invokeCommand<AgentResourceLock[]>("research_resource_lock_list", {
    input: { projectId },
  });
}

export async function resolveResearchPlanApproval(
  projectId: string,
  approvalId: string,
  decision: "approved" | "rejected",
): Promise<ResearchPlanExecutionAccepted> {
  const accepted = await invokeCommand<ResearchPlanExecutionAccepted>("research_plan_approval_resolve", {
    input: { projectId, approvalId, decision },
  });
  emitResearchRunChanged(projectId);
  return accepted;
}

export function listResearchEvidence(projectId: string, taskId: string): Promise<EvidencePacket[]> {
  return invokeCommand<EvidencePacket[]>("research_evidence_list", {
    input: { projectId, taskId },
  });
}

export function assessResearchClaim(input: {
  projectId: string;
  taskId: string;
  claim: string;
  evidenceIds: string[];
  repairedClaim?: string | null;
}): Promise<ClaimEvidenceAssessment> {
  return invokeCommand<ClaimEvidenceAssessment>("research_claim_assess", { input });
}

export function listResearchClaimAssessments(
  projectId: string,
  taskId: string,
): Promise<ClaimEvidenceAssessment[]> {
  return invokeCommand<ClaimEvidenceAssessment[]>("research_claim_assessment_list", {
    input: { projectId, taskId },
  });
}

export function getResearchFulltextDocument(
  projectId: string,
  documentHash: string,
): Promise<ResearchFulltextDocument> {
  return invokeCommand<ResearchFulltextDocument>("research_fulltext_document_get", {
    input: { projectId, documentHash },
  });
}

export function saveResearchReviewProtocol(input: {
  projectId: string;
  taskId: string;
  title: string;
  researchQuestion: string;
  inclusionCriteria: string[];
  exclusionCriteria: string[];
}): Promise<ResearchReviewProtocol> {
  return invokeCommand<ResearchReviewProtocol>("research_review_protocol_save", { input });
}

export function recordResearchQuerySnapshot(input: {
  projectId: string;
  taskId: string;
  stableId?: string | null;
  query: string;
  sources: string[];
  resultCount: number;
  stopReason: ResearchQuerySnapshot["stopReason"];
}): Promise<ResearchQuerySnapshot> {
  return invokeCommand<ResearchQuerySnapshot>("research_review_query_snapshot_record", { input });
}

export function suggestResearchScreening(input: {
  projectId: string;
  taskId: string;
  evidenceId: string;
  recommendation: ResearchScreeningRecord["recommendation"];
  confidence: number;
  suggestionReason: string;
}): Promise<ResearchScreeningRecord> {
  return invokeCommand<ResearchScreeningRecord>("research_review_screening_suggest", { input });
}

export function confirmResearchScreenings(input: {
  projectId: string;
  taskId: string;
  decisions: Array<{
    screeningId: string;
    decision: "include" | "exclude";
    exclusionReason?: string | null;
    fullTextReviewed: boolean;
  }>;
}): Promise<ResearchReviewWorkspace> {
  return invokeCommand<ResearchReviewWorkspace>("research_review_screening_confirm_batch", { input });
}

export function getResearchReviewWorkspace(
  projectId: string,
  taskId: string,
): Promise<ResearchReviewWorkspace> {
  return invokeCommand<ResearchReviewWorkspace>("research_review_workspace_get", {
    input: { projectId, taskId },
  });
}

File: src-tauri/src/commands/research_agent.rs
use crate::models::{
    Ack, AgentResourceLock, AgentResourceLockListInput, AgentResourceLockReleaseInput,
    ClaimEvidenceAssessInput, ClaimEvidenceAssessment, ClaimEvidenceAssessmentListInput,
    EvidencePacket, EvidencePacketListInput, EvidencePacketUpsertInput, ResearchAgentRun,
    ResearchCapabilityDescriptor, ResearchChangeCheckpoint, ResearchChangeCheckpointListInput,
    ResearchChangeCheckpointUndoInput, ResearchChangeCheckpointUndoResult,
    ResearchChatMigrationInput, ResearchChatMigrationResult, ResearchChatStore,
    ResearchChatStoreReplaceInput, ResearchPlanApproval, ResearchPlanApprovalResolveInput,
    ResearchPlanApproveInput, ResearchPlanExecuteInput, ResearchPlanExecutionAccepted,
    ResearchPlanSaveInput, ResearchPlanVersion, ResearchProjectInput, ResearchRunControlInput,
    ResearchRunListInput, ResearchRunRecoveryResponse, ResearchRunsRecoverInput, ResearchTask,
    ResearchTaskCreateInput, ResearchUiCommand, ResearchUiCommandListInput,
    ResearchUiCommandResolveInput, ResearchWorkspaceSnapshot,
};
use crate::state::AppState;
use crate::storage;
use tauri::State;

#[tauri::command]
pub fn research_capability_registry() -> Vec<ResearchCapabilityDescriptor> {
    crate::research_agent::capability_registry()
}

#[tauri::command]
pub fn research_workspace_get(
    state: State<'_, AppState>,
    input: ResearchProjectInput,
) -> Result<ResearchWorkspaceSnapshot, String> {
    storage::research_workspace_snapshot(&state.db_path, &state.runtime_root, &input.project_id)
}

#[tauri::command]
pub fn research_task_create(
    state: State<'_, AppState>,
    input: ResearchTaskCreateInput,
) -> Result<ResearchTask, String> {
    storage::create_research_task(&state.db_path, &state.runtime_root, input)
}

#[tauri::command]
pub fn research_plan_save(
    state: State<'_, AppState>,
    input: ResearchPlanSaveInput,
) -> Result<ResearchPlanVersion, String> {
    storage::save_research_plan(&state.db_path, &state.runtime_root, input)
}

#[tauri::command]
pub fn research_plan_approve(
    state: State<'_, AppState>,
    input: ResearchPlanApproveInput,
) -> Result<ResearchPlanVersion, String> {
    storage::approve_research_plan(&state.db_path, &state.runtime_root, input)
}

#[tauri::command]
pub fn research_chat_store_get(
    state: State<'_, AppState>,
    input: ResearchProjectInput,
) -> Result<ResearchChatStore, String> {
    storage::research_chat_store_get(&state.db_path, &state.runtime_root, &input.project_id)
}

#[tauri::command]
pub fn research_chat_store_replace(
    state: State<'_, AppState>,
    input: ResearchChatStoreReplaceInput,
) -> Result<ResearchChatStore, String> {
    storage::research_chat_store_replace(&state.db_path, &state.runtime_root, input)
}

#[tauri::command]
pub fn research_chat_store_migrate(
    state: State<'_, AppState>,
    input: ResearchChatMigrationInput,
) -> Result<ResearchChatMigrationResult, String> {
    storage::research_chat_store_migrate(&state.db_path, &state.runtime_root, input)
}

#[tauri::command]
pub fn research_resource_lock_list(
    state: State<'_, AppState>,
    input: AgentResourceLockListInput,
) -> Result<Vec<AgentResourceLock>, String> {
    storage::list_research_resource_locks(&state.db_path, &input.project_id)
}

#[tauri::command]
pub fn research_resource_lock_release(
    state: State<'_, AppState>,
    input: AgentResourceLockReleaseInput,
) -> Result<Ack, String> {
    let released =
        storage::release_research_resource_locks(&state.db_path, &input.project_id, &input.run_id)?;
    Ok(Ack {
        ok: true,
        message: format!("research.lock.released:{released}"),
    })
}

#[tauri::command]
pub fn research_plan_execute(
    state: State<'_, AppState>,
    input: ResearchPlanExecuteInput,
) -> Result<ResearchPlanExecutionAccepted, String> {
    super::research_agent_execution::start_plan_execution(
        &state,
        &input.project_id,
        &input.task_id,
        input.version,
    )
}

#[tauri::command]
pub fn research_run_list(
    state: State<'_, AppState>,
    input: ResearchRunListInput,
) -> Result<Vec<ResearchAgentRun>, String> {
    let project_ids = if let Some(project_id) = input.project_id {
        vec![project_id]
    } else {
        storage::list_projects(&state.db_path)?
            .into_iter()
            .map(|project| project.id)
            .collect()
    };
    let mut runs = Vec::new();
    for project_id in project_ids {
        runs.extend(storage::list_research_plan_runs(
            &state.db_path,
            &state.runtime_root,
            &project_id,
            input.include_terminal.unwrap_or(false),
        )?);
    }
    runs.sort_by(|left, right| right.updated_at.cmp(&left.updated_at));
    Ok(runs)
}

#[tauri::command]
pub fn research_ui_command_list(
    state: State<'_, AppState>,
    input: ResearchUiCommandListInput,
) -> Result<Vec<ResearchUiCommand>, String> {
    storage::list_pending_research_ui_commands(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
    )
}

#[tauri::command]
pub fn research_ui_command_resolve(
    state: State<'_, AppState>,
    input: ResearchUiCommandResolveInput,
) -> Result<ResearchPlanExecutionAccepted, String> {
    let pending = storage::list_pending_research_ui_commands(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
    )?
    .into_iter()
    .find(|command| command.run_id == input.run_id && command.step_id == input.step_id)
    .ok_or_else(|| "research.ui_command.not_pending".to_string())?;
    if input.status == "completed"
        && matches!(
            pending.command,
            crate::models::AgentAppCommand::ApplyLatexProposal { .. }
        )
    {
        storage::finalize_research_change_checkpoint(
            &state.db_path,
            &state.runtime_root,
            &input.project_id,
            &input.run_id,
            &input.step_id,
        )?;
    }
    let completed_steps =
        storage::resolve_research_ui_command(&state.db_path, &state.runtime_root, &input)?;
    storage::release_research_resource_locks(&state.db_path, &input.project_id, &input.run_id)?;
    if input.status == "completed" {
        storage::update_research_run_progress(
            &state.db_path,
            &state.runtime_root,
            &input.project_id,
            &input.run_id,
            "waiting_ui",
            Some(&input.step_id),
            completed_steps,
            Some("ui_command_completed"),
            None,
        )?;
        return super::research_agent_execution::resume_plan_execution(
            &state,
            &input.project_id,
            &input.run_id,
        );
    }
    storage::update_research_run_progress(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
        &input.run_id,
        "failed",
        Some(&input.step_id),
        completed_steps,
        Some("ui_command_failed"),
        input.diagnostic_code.as_deref(),
    )?;
    Ok(ResearchPlanExecutionAccepted {
        run_id: input.run_id,
        status: "failed".to_string(),
    })
}

fn set_research_run_status(
    state: &AppState,
    input: &ResearchRunControlInput,
    status: &str,
) -> Result<ResearchPlanExecutionAccepted, String> {
    let run = storage::get_research_plan_run(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
        &input.run_id,
    )?;
    storage::update_research_run_progress(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
        &input.run_id,
        status,
        run.current_step_id.as_deref(),
        run.completed_steps,
        run.last_operation.as_deref(),
        None,
    )?;
    if status == "cancelled" {
        if !storage::research_run_has_active_lease(
            &state.db_path,
            &input.project_id,
            &input.run_id,
        )? {
            storage::release_research_resource_locks(
                &state.db_path,
                &input.project_id,
                &input.run_id,
            )?;
            storage::cancel_research_run_lease(&state.db_path, &input.project_id, &input.run_id)?;
        }
    }
    Ok(ResearchPlanExecutionAccepted {
        run_id: input.run_id.clone(),
        status: status.to_string(),
    })
}

#[tauri::command]
pub fn research_run_pause(
    state: State<'_, AppState>,
    input: ResearchRunControlInput,
) -> Result<ResearchPlanExecutionAccepted, String> {
    set_research_run_status(&state, &input, "paused")
}

#[tauri::command]
pub fn research_run_cancel(
    state: State<'_, AppState>,
    input: ResearchRunControlInput,
) -> Result<ResearchPlanExecutionAccepted, String> {
    set_research_run_status(&state, &input, "cancelled")
}

#[tauri::command]
pub fn research_run_resume(
    state: State<'_, AppState>,
    input: ResearchRunControlInput,
) -> Result<ResearchPlanExecutionAccepted, String> {
    super::research_agent_execution::resume_plan_execution(&state, &input.project_id, &input.run_id)
}

#[tauri::command]
pub fn research_runs_recover(
    state: State<'_, AppState>,
    input: ResearchRunsRecoverInput,
) -> Result<ResearchRunRecoveryResponse, String> {
    super::research_agent_execution::recover_plan_executions(&state, &input.project_id)
}

#[tauri::command]
pub fn research_change_checkpoint_list(
    state: State<'_, AppState>,
    input: ResearchChangeCheckpointListInput,
) -> Result<Vec<ResearchChangeCheckpoint>, String> {
    storage::list_research_change_checkpoints(
        &state.db_path,
        &input.project_id,
        input.run_id.as_deref(),
    )
}

#[tauri::command]
pub fn research_change_checkpoint_undo(
    state: State<'_, AppState>,
    input: ResearchChangeCheckpointUndoInput,
) -> Result<ResearchChangeCheckpointUndoResult, String> {
    storage::undo_research_change_checkpoint(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
        &input.checkpoint_id,
    )
}

#[tauri::command]
pub fn research_plan_approval_list(
    state: State<'_, AppState>,
    input: ResearchProjectInput,
) -> Result<Vec<ResearchPlanApproval>, String> {
    storage::list_research_plan_approvals(&state.db_path, &state.runtime_root, &input.project_id)
}

#[tauri::command]
pub fn research_plan_approval_resolve(
    state: State<'_, AppState>,
    input: ResearchPlanApprovalResolveInput,
) -> Result<ResearchPlanExecutionAccepted, String> {
    let (run_id, _) = storage::resolve_research_plan_approval(
        &state.db_path,
        &input.project_id,
        &input.approval_id,
        &input.decision,
    )?;
    if input.decision == "approved" {
        return super::research_agent_execution::resume_plan_execution(
            &state,
            &input.project_id,
            &run_id,
        );
    }
    set_research_run_status(
        &state,
        &ResearchRunControlInput {
            project_id: input.project_id,
            run_id,
        },
        "cancelled",
    )
}

#[tauri::command]
pub fn research_evidence_upsert(
    state: State<'_, AppState>,
    input: EvidencePacketUpsertInput,
) -> Result<EvidencePacket, String> {
    storage::upsert_evidence_packet(&state.db_path, &state.runtime_root, input)
}

#[tauri::command]
pub fn research_evidence_list(
    state: State<'_, AppState>,
    input: EvidencePacketListInput,
) -> Result<Vec<EvidencePacket>, String> {
    storage::list_evidence_packets(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
        &input.task_id,
    )
}

#[tauri::command]
pub fn research_claim_assess(
    state: State<'_, AppState>,
    input: ClaimEvidenceAssessInput,
) -> Result<ClaimEvidenceAssessment, String> {
    storage::assess_claim_evidence(&state.db_path, &state.runtime_root, input)
}

#[tauri::command]
pub fn research_claim_assessment_list(
    state: State<'_, AppState>,
    input: ClaimEvidenceAssessmentListInput,
) -> Result<Vec<ClaimEvidenceAssessment>, String> {
    storage::list_claim_evidence_assessments(
        &state.db_path,
        &state.runtime_root,
        &input.project_id,
        &input.task_id,
    )
}
